File: typescript/packages/bench/src/contracts/compositeContract.ts
/** Interface with three parents: exercises multi-parent fan-out. */

import type { CacheableContract } from "./cacheableContract";
import type { ReportContract } from "./reportContract";
import type { ScheduleContract } from "./scheduleContract";

export interface CompositeContract extends ReportContract, CacheableContract, ScheduleContract {}

File: python/atelier_bench/contracts/composite_contract.py
"""Multi-parent surface: three parent contracts in one class statement (real multiple inheritance)."""

from __future__ import annotations

from .cacheable_contract import CacheableContract
from .report_contract import ReportContract
from .schedule_contract import ScheduleContract


class CompositeContract(ReportContract, CacheableContract, ScheduleContract):
    """Abstract (inherits ReportContract.rows); the MRO carries all three parents."""

    def parents(self) -> tuple[str, ...]:
        """The three direct bases of CompositeContract itself, whichever concrete class asks."""
        return tuple(base.__name__ for base in CompositeContract.__bases__)
